### PR TITLE
chore(starknet_l1_provider): implement event fetch

### PR DESCRIPTION
**Pull request by giladchase**

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2854)* status: all files reviewed, 5 unresolved discussions (waiting on @giladchase and @matan-starkware)

___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 57 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsxcNz4tP0FuShT4Op:-OEsxcNz4tP0FuShT4Oq:b-d31bcn) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L57)):*
> ```XML
>             error!("Failed to get latest L1 block number, finality too high.");
>             return Ok(());
>         };
> ```

In case this code is repeated, consider a function.



_Code quote:_
```XML
        let latest_l1_block_number = self
            .base_layer
            .latest_l1_block_number(self.config.finality)
            .await
            .map_err(L1ScraperError::BaseLayer)?;

        let Some(latest_l1_block_number) = latest_l1_block_number else {
            error!("Failed to get latest L1 block number, finality too high.");
            return Ok(());
        };
```
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 72 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsxsFhCRK-Su2tbbUo:-OEsxsFhCRK-Su2tbbUp:b-6g1ax1) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L72)):*
> ```XML
>             .into_iter()
>             .map(|event| self.event_from_raw_l1_event(event))
>             .collect::<L1ScraperResult<Vec<_>, _>>()?;
> ```

Do you prefer turbofish over type annotations?
Also, WDYM by `collect`ing to a `Result`?



_Code quote:_
```XML
::<L1ScraperResult<Vec<_>, _>>
```
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 86 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsyKD6DElHKq-P9mT8:-OEsyKD6DElHKq-P9mT9:ba75ln8) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L86)):*
> ```XML
>     }
> 
>     fn event_from_raw_l1_event(&self, l1_event: L1Event) -> L1ScraperResult<Event, B> {
> ```

Can this be `Event::TryFrom<L1Event>` method?



_Code quote:_
```XML
event_from_raw_l1_event
```
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 92 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsz0H50WJfP-p15zYL:-OEsz0H50WJfP-p15zYM:b-n69ihl) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L92)):*
> ```XML
>                 ExecutableL1HandlerTransaction::create(tx, chain_id, fee)
>                     .map(Event::L1HandlerTransaction)
>                     .map_err(L1ScraperError::HashCalculationError)
> ```

Not very readable, IMO. Any alternative?



_Code quote:_
```XML
                ExecutableL1HandlerTransaction::create(tx, chain_id, fee)
                    .map(Event::L1HandlerTransaction)
                    .map_err(L1ScraperError::HashCalculationError)
```
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 124 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEszEF75us3-OeNTSJf:-OEszEF75us3-OeNTSJg:bgny4eq) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L124)):*
> ```XML
> 
> #[derive(Error, Debug)]
> pub enum L1ScraperError<T: BaseLayerContract + Send + Sync> {
> ```

Is this necessary? What exactly is generic here - the L1, or the contract we're interacting with?



_Code quote:_
```XML
T: BaseLayerContract
```

**Review by matan-starkware — APPROVED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2854)* status: all files reviewed, 8 unresolved discussions (waiting on @elintul and @giladchase)

___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 72 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsxsFhCRK-Su2tbbUo:-OExDI_s4fmVHl1xHREX:b-p6b7z0) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L72)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Do you prefer turbofish over type annotations?
Also, WDYM by `collect`ing to a `Result`?
</blockquote></details>

"Also, WDYM by `collect`ing to a `Result`?" This tells Rust to iterate over `Result<T>` elements and upon encountering a single error exit, returning that error. If no errors are encountered, it produces `Vec<T>`.
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 92 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsz0H50WJfP-p15zYL:-OExEnlxCGd3z6JytRCF:b7tvnvn) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L92)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Not very readable, IMO. Any alternative?
</blockquote></details>

Agreed. If you update the error type to `    HashCalculationError(#[from] StarknetApiError),` Then you could do
```
                let tx = ExecutableL1HandlerTransaction::create(tx, &self.config.chain_id, fee)?;
                Ok(Event::L1HandlerTransaction(tx))
```
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 55 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OExCKYF7UvvRPdOpFyz:-OExCKYG18EBNaNdfoN6:be4gcyh) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L55)):*
> ```XML
> 
>         let Some(latest_l1_block_number) = latest_l1_block_number else {
>             error!("Failed to get latest L1 block number, finality too high.");
> ```

This means that there is no finalized block on ethereum? (dependent upon our config)



_Code quote:_
```XML
            error!("Failed to get latest L1 block number, finality too high.");
```
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 74 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OExDvBn6x875FCSJ1SB:-OExDvBn6x875FCSJ1SC:b5nx1sp) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L74)):*
> ```XML
>             .collect::<L1ScraperResult<Vec<_>, _>>()?;
> 
>         self.last_block_number_processed = latest_l1_block_number + 1;
> ```

Is `BaseLayerContract::until_block` inclusive on `until_block`? I think that should be explicit in the trait doc.



_Code quote:_
```XML
        self.last_block_number_processed = latest_l1_block_number + 1;
```
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 128 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OExFHCJ4Y-9-vBrwl_k:-OExFHCJ4Y-9-vBrwl_l:b-q9ibse) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L128)):*
> ```XML
>     BaseLayer(T::Error),
>     #[error("Failed to calculate hash: {0}")]
>     HashCalculationError(StarknetApiError),
> ```

Allow for easy conversions



_Suggestion:_
```XML
    HashCalculationError(#[from] StarknetApiError),
```

**Review by elintul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2854)* status: all files reviewed, 8 unresolved discussions (waiting on @giladchase and @matan-starkware)

___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 72 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsxsFhCRK-Su2tbbUo:-OF0vOTk4iGAUFMYVb8u:b-ucfknh) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L72)):*
<details><summary><i>Previously, matan-starkware wrote…</i></summary><blockquote>

"Also, WDYM by `collect`ing to a `Result`?" This tells Rust to iterate over `Result<T>` elements and upon encountering a single error exit, returning that error. If no errors are encountered, it produces `Vec<T>`.
</blockquote></details>

Thanks!

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2854)* status: all files reviewed, 7 unresolved discussions (waiting on @elintul and @matan-starkware)

___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 55 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OExCKYF7UvvRPdOpFyz:-OExqljz6PXznAW8QACa:b-i39v9) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L55)):*
<details><summary><i>Previously, matan-starkware wrote…</i></summary><blockquote>

This means that there is no finalized block on ethereum? (dependent upon our config)
</blockquote></details>

Yep. I didn't want to throw an error, because there are cases where this can happen (like in genesis).
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 57 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsxcNz4tP0FuShT4Op:-OExrTaTCPVWXCJSP2vP:b-zh3qfq) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L57)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

In case this code is repeated, consider a function.
</blockquote></details>

I don't think i'll need this again in the scraper, unless I'm misunderstanding what you mean here
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 74 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OExDvBn6x875FCSJ1SB:-OFv9ErO-W9FmDbQwG7x:b-6edec3) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L74)):*
<details><summary><i>Previously, matan-starkware wrote…</i></summary><blockquote>

Is `BaseLayerContract::until_block` inclusive on `until_block`? I think that should be explicit in the trait doc.
</blockquote></details>

Good point.
Documenting it with types if that's OK :muscle: will now take `RangeInclusive` as an arg
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 86 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsyKD6DElHKq-P9mT8:-OFv9NVtEyjIQ4m8DcTH:b-t9grsk) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L86)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Can this be `Event::TryFrom<L1Event>` method?
</blockquote></details>

Nope it needs chain_id for l1Handler variant :(
I though about maybe doing a try_from then add the chain afterwards, but this won't work cause I need the chain_id to calculate the tx_hash :smiling_face_with_tear: 
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 92 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsz0H50WJfP-p15zYL:-OFvBRKb-CJ_oq9RGNhE:b-5hnbfh) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L92)):*
> Any alternative?

How about this?
```rust
            L1Event::LogMessageToL2 { tx, fee } => {
                let chain_id = &self.config.chain_id;
                match ExecutableL1HandlerTransaction::create(tx, chain_id, fee) {
                    Ok(tx) => Ok(Event::L1HandlerTransaction(tx)),
                    Err(hash_calc_err) => Err(L1ScraperError::HashCalculationError(hash_calc_err)),
                }
            }
```

>If you update the error type to `HashCalculationError(#[from] StarknetApiError)`

I wanted to avoid doing that since this hash calc is the only SNAPI error possible here, whereas `#[from]`  catches all variants.

We've already hit this issue in the past, where future devs of such code used `?` on StarknetApi because "it just worked", and later on people suddenly got errors like: `HashCalcError("<some error that isn't related to hash calc>")`.

In other words, if we add `#[from]`, we should change the error to StarknetApiError and lose the specificity of HashCalcError variant, which in turn makes future readers of this code unable to distinguish which error variants are actually possible here (which currently is just the one).

Personally, I tend now only to use `#[from]` if at least `min(3,n_variants_of_error)` variants are possible, but of course, **if you still prefer using `#[from]`, I'll change it**.
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 124 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEszEF75us3-OeNTSJf:-OFv9pCy1gC1VgN3T1EL:b-45c577) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L124)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Is this necessary? What exactly is generic here - the L1, or the contract we're interacting with?
</blockquote></details>

Kind of both. It is assumed to be some wrapper around a crate that interacts with L1, that is initialized with _the_ hardcoded Starknet contract on L1.

This was useful for testing, I could mock it easily.
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 128 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OExFHCJ4Y-9-vBrwl_k:-OFvDkR1-1v_K0K87DXf:bxi9yam) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L128)):*
<details><summary><i>Previously, matan-starkware wrote…</i></summary><blockquote>

Allow for easy conversions
</blockquote></details>

answered [in this thread](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsz0H50WJfP-p15zYL)

**Review by matan-starkware — APPROVED** [review]
> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2854)* status: all files reviewed, 5 unresolved discussions (waiting on @elintul and @giladchase)

___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 74 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OExDvBn6x875FCSJ1SB:-OFvcBMO-643YXzY2B-P:b33w1d7) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L74)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Good point.
Documenting it with types if that's OK :muscle: will now take `RangeInclusive` as an arg
</blockquote></details>

Yeah the new interface is nice, and makes the code inside shorter also
___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 92 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsz0H50WJfP-p15zYL:-OFvcq4C2CII_3fk4cNH:bb73djm) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L92)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

> Any alternative?

How about this?
```rust
            L1Event::LogMessageToL2 { tx, fee } => {
                let chain_id = &self.config.chain_id;
                match ExecutableL1HandlerTransaction::create(tx, chain_id, fee) {
                    Ok(tx) => Ok(Event::L1HandlerTransaction(tx)),
                    Err(hash_calc_err) => Err(L1ScraperError::HashCalculationError(hash_calc_err)),
                }
            }
```

>If you update the error type to `HashCalculationError(#[from] StarknetApiError)`

I wanted to avoid doing that since this hash calc is the only SNAPI error possible here, whereas `#[from]`  catches all variants.

We've already hit this issue in the past, where future devs of such code used `?` on StarknetApi because "it just worked", and later on people suddenly got errors like: `HashCalcError("<some error that isn't related to hash calc>")`.

In other words, if we add `#[from]`, we should change the error to StarknetApiError and lose the specificity of HashCalcError variant, which in turn makes future readers of this code unable to distinguish which error variants are actually possible here (which currently is just the one).

Personally, I tend now only to use `#[from]` if at least `min(3,n_variants_of_error)` variants are possible, but of course, **if you still prefer using `#[from]`, I'll change it**.
</blockquote></details>

SGTM

**Review by giladchase — COMMENTED** [review]
> Dismissed @elintul from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2854)* status: 6 of 7 files reviewed, 2 unresolved discussions (waiting on @elintul and @matan-starkware)

___
*[`crates/starknet_l1_provider/src/l1_scraper.rs` line 92 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2854#-OEsz0H50WJfP-p15zYL:-OFvt9o4Dhze3PKmW1BL:b-97tttv) ([raw file](https://github.com/starkware-libs/sequencer/blob/a840e251393674f486bac99dc151a31764d13ffa/crates/starknet_l1_provider/src/l1_scraper.rs#L92)):*
<details><summary><i>Previously, matan-starkware wrote…</i></summary><blockquote>

SGTM
</blockquote></details>

Done. 
Elin is OOO so i'm gonna dismiss the block and merge ahead unless there are objections :grimacing:

**Review by giladchase — COMMENTED** [review]
> Reviewed 2 of 5 files at r1, 4 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2854)* status: all files reviewed, 2 unresolved discussions (waiting on @elintul)
